src/GraphQL/Validation/Rules/NoFragmentCycles.cs

```diff
@@ -20,36 +20,28 @@ public class NoFragmentCycles : IValidationRule
 
         /// <inheritdoc/>
         /// <exception cref="NoFragmentCyclesError"/>
-        public Task<INodeVisitor> ValidateAsync(ValidationContext context)
-        {
-            // Tracks already visited fragments to maintain O(N) and to ensure that cycles
-            // are not redundantly reported.
-            var visitedFrags = new LightweightCache<string, bool>(key => false);
-
-            // Array of AST nodes used to produce meaningful errors
-            var spreadPath = new Stack<FragmentSpread>();
-
-            // Position in the spread path
-            var spreadPathIndexByName = new LightweightCache<string, int>(key => -1);
+        public Task<INodeVisitor> ValidateAsync(ValidationContext context) => context.Document.Fragments.Count > 0 ? _nodeVisitor : null;
```

Comment:
Your short-circuit algorithm is implemented here 👍 .  Returns null if no need to run the visitor.